datafusion/expr/src/expr_schema.rs

```diff
@@ -107,7 +105,7 @@ impl ExprSchemable for Expr {
                     .iter()
                     .map(|e| e.get_type(schema))
                     .collect::<Result<Vec<_>>>()?;
-                function::return_type(fun, &data_types)
+                fun.return_type(&data_types)
```

Comment:
this is a pretty good example of the point of this PR -- the functionality is now part of `fun` rather than a free function that takes `&self` as its first argument

Reply:
(I hope to apply the same pattern to `window_function::return_type` and `aggregate_function::return_type`

Reply:
>  (I hope to apply the same pattern to window_function::return_type and aggregate_function::return_type

🚀agree with it